tracer/src/Datadog.Trace/ClrProfiler/CallTarget/CallTargetInvoker.cs

```diff
@@ -18,13 +20,41 @@ namespace Datadog.Trace.ClrProfiler.CallTarget;
 [EditorBrowsable(EditorBrowsableState.Never)]
 public static class CallTargetInvoker
 {
+#if NETFRAMEWORK
+    private const string NamedSlotName = "Datadog_IISPreInitStart";
+    private static bool _isIisPreStartInitComplete = false;
+#endif
+
     static CallTargetInvoker()
     {
         // The first time the CallTargetInvoker is called
         // we ensure that the non native parts of the initialization ran
         // This is required for AOT scenarios where there is no clrprofiler
         // to inject and run the loader.
-        Instrumentation.InitializeNoNativeParts();
+        // TODO: We don't support AOT scenarios yet, so this is disabled for now
+        // as it can cause issues by initializing the profiler too early
+        // (during IIS pre-init)
+        // Instrumentation.InitializeNoNativeParts();
+
+#if NETFRAMEWORK
+        // Check if IIS automatic instrumentation has set the AppDomain property to indicate the PreStartInit state
+        // If the property is not set, we should rely on other heuristics
+        var state = AppDomain.CurrentDomain.GetData(NamedSlotName);
+        if (state is bool boolState)
+        {
+            _isIisPreStartInitComplete = !boolState;
+        }
+        else
+        {
+            // This _theoretically_ is a problem. In previous workarounds
```

Comment:
The code comment says this branch "theoretically is a problem" but we generally expect to fall in this branch for any non-IIS .NET Framework application, right? What is the scenario in which it would be a problem? Is it if we should be inserting the state but we do not?

The reason I'm adding this comment is because the first sentence about being a problem is jarring and then I realize that this is actually fine in most cases 😅

Reply:
> The code comment says this branch "theoretically is a problem" but we generally expect to fall in this branch for any non-IIS .NET Framework application, right? What is the scenario in which it would be a problem? Is it if we should be inserting the state but we do not?

Yeah, I think non-IIS is the most likely source, but in [your previous implementation](https://github.com/DataDog/dd-trace-dotnet/pull/1157) you went to great pains to check the process name, and to read the frames from the stack... So the "theoretically" here is because I YOLO'd it, as I thought that seemed overkill 😅 

Unfortuntately, this seems to completely kill the secondary app pool, and due to a _different_ bug, I'm struggling to figure out why 😅

Reply:
So, tl;dr; the `else` was completely wrong, so this documentation has all changed 🙈